SU2_CFD/src/solvers/CSpeciesSolver.cpp

```diff
@@ -0,0 +1,529 @@
+  auto SolverSpecificNumerics = [&](unsigned long iPoint, unsigned long jPoint) {
+    /*--- Mass diffusivity coefficients. ---*/
+
+    numerics->SetDiffusionCoeff(nodes->GetDiffusivity(iPoint), nodes->GetDiffusivity(jPoint));
+  };
+
+  /*--- Now instantiate the generic implementation with the functor above. ---*/
+
+  Viscous_Residual_impl(SolverSpecificNumerics, iEdge, geometry, solver_container, numerics, config);
+}
+
+void CSpeciesSolver::BC_Inlet(CGeometry* geometry, CSolver** solver_container, CNumerics* conv_numerics,
+                              CNumerics* visc_numerics, CConfig* config, unsigned short val_marker) {
+  /// NOTE TK:: This is a strong impl whereas TurbSA and inceuler implement a weak version. Testing required.
+  su2double* inlet_species = new su2double[nVar]{};
+
+  // bool grid_movement  = config->GetGrid_Movement();
+  string Marker_Tag = config->GetMarker_All_TagBound(val_marker);
+  // su2double   temp_inlet    = config->GetInlet_Ttotal       (Marker_Tag);
+  inlet_species = config->GetInlet_SpeciesVal(Marker_Tag);
+
+  /*--- Loop over all the vertices on this boundary marker ---*/
+
+  SU2_OMP_FOR_STAT(OMP_MIN_SIZE)
+  for (auto iVertex = 0u; iVertex < geometry->nVertex[val_marker]; iVertex++) {
+    auto iPoint = geometry->vertex[val_marker][iVertex]->GetNode();
+
+    /*--- Check if the node belongs to the domain (i.e., not a halo node) ---*/
+
+    if (geometry->nodes->GetDomain(iPoint)) {
+      nodes->SetSolution_Old(iPoint, inlet_species);
+
+      LinSysRes.SetBlock_Zero(iPoint);
+
+      for (auto iVar = 0u; iVar < nVar; iVar++) {
+        nodes->SetVal_ResTruncError_Zero(iPoint, iVar);
+      }
+
+      /*--- Includes 1 in the diagonal ---*/
+      for (auto iVar = 0u; iVar < nVar; iVar++) {
+        auto total_index = iPoint * nVar + iVar;
+        Jacobian.DeleteValsRowi(total_index);
+      }
+    }
+  }
+  END_SU2_OMP_FOR
+}
+
+void CSpeciesSolver::BC_Outlet(CGeometry* geometry, CSolver** solver_container, CNumerics* conv_numerics,
+                               CNumerics* visc_numerics, CConfig* config, unsigned short val_marker) {
+  /// NOTE TK:: This is a strong impl whereas TurbSA and inceuler implement a weak version. Testing required.
+  /*--- Loop over all the vertices on this boundary marker ---*/
+
+  SU2_OMP_FOR_STAT(OMP_MIN_SIZE)
+  for (auto iVertex = 0u; iVertex < geometry->nVertex[val_marker]; iVertex++) {
+    /* strong zero flux Neumann boundary condition at the outlet */
+    const auto iPoint = geometry->vertex[val_marker][iVertex]->GetNode();
+
+    /*--- Check if the node belongs to the domain (i.e., not a halo node) ---*/
+
+    if (geometry->nodes->GetDomain(iPoint)) {
+      /*--- Allocate the value at the outlet ---*/
+      auto Point_Normal = geometry->vertex[val_marker][iVertex]->GetNormal_Neighbor();
+
+      for (auto iVar = 0u; iVar < nVar; iVar++) Solution[iVar] = nodes->GetSolution(Point_Normal, iVar);
+      nodes->SetSolution_Old(iPoint, Solution);
```

Comment:
only works on orthogonal grids
```suggestion
      nodes->SetSolution_Old(iPoint, nodes->GetSolution(Point_Normal));
```

Reply:
this change was applied manually at some point